Add Relationships Iterator and Sort Methods in Header

s9latimm opened the pull request.

s9latimm committed:
Add iterator and sort methods

M	include/vara/Feature/FeatureModel.h

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -240,19 +248,44 @@ class FeatureModel {
+  Constraint *addConstraint(std::unique_ptr<Constraint> Constraint) {
+    Constraints.push_back(std::move(Constraint));
+    return Constraints.back().get();
+  }
```
Should not be public. This does against the idea we implemented with feature transactions. By this users are now able to add constraints that create an invalid feature model. Why would we need this change?

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -240,19 +248,44 @@ class FeatureModel {
+    return llvm::make_range(Relationships.begin(), Relationships.end());
+  }
+
+  Relationship *addRelationship(std::unique_ptr<Relationship> Relationship) {
+    Relationships.push_back(std::move(Relationship));
+    return Relationships.back().get();
+  }
```
Same here, users can now freely edit relationships.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -240,19 +248,44 @@ class FeatureModel {
+    return Relationships.back().get();
+  }
+
+  void removeRelationship(Relationship *R) {
```
This one should be fine as, for what I know, removing a relationship cannot create an invalid model from a previously valid one.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -240,19 +248,44 @@ class FeatureModel {
-      return SearchFeature->getValue().get();
-    }
-    return nullptr;
+  [[nodiscard]] llvm::iterator_range<relationship_iterator> relationships() {
+    return llvm::make_range(Relationships.begin(), Relationships.end());
+  }
```
I'm not sure if exposing non const iterators serves our purpose. Wouldn't his also allow modifications on the user side?

s9latimm committed:
Add const

M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-RelationshipIterator

M	.github/workflows/build.yml
M	.github/workflows/deploydocs.yml
M	.github/workflows/reviewdog.yml
M	CMakeLists.txt
M	README.md
M	external/pybind11
M	include/vara/Feature/ConstraintParser.h
M	include/vara/Feature/FeatureModel.h
M	scripts/ci_utils/prepare_coverage_data.py

s9latimm committed:
No need for const here

M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Fix ordering

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Fix bindings

M	bindings/python/vara-feature/pybind_FeatureModel.cpp
